src/components/Menu/ControlledMenu.js

```diff
@@ -0,0 +1,174 @@
+   */
+  renderClosedDropdown: PropTypes.bool,
+};
+
+class ControlledMenu extends Component {
+  static propTypes = propTypes;
+
+  static defaultProps = {
+    className: null,
+    checkbox: false,
+    isOpen: false,
+    last: false,
+    nubbin: false,
+    position: 'top-left',
+    size: 'small',
+    height: null,
+    heightIcon: null,
+    renderClosedDropdown: false,
+  };
+
+  getClasses = () => {
+    const { isOpen, last } = this.props;
+
+    const classes = [
+      'slds-dropdown-trigger',
+      'slds-dropdown-trigger_click',
+      { 'slds-button_last': last },
+    ];
+    if (!isOpen) {
+      return classes;
+    }
+    return [...classes, 'slds-is-open'];
+  };
+
+  getDropdownClasses = () => {
+    const {
+      size,
+      position,
+      className,
+      nubbin
+    } = this.props;
+
+    const dropdownClasses = [
+      'slds-dropdown',
+      { [`slds-dropdown_${size}`]: size },
+      { 'slds-dropdown_left': position.endsWith('left') },
+      { 'slds-dropdown_right': position.endsWith('right') },
+      { 'slds-dropdown_bottom': position.startsWith('bottom') },
+      { [`slds-nubbin_${position}`]: nubbin },
+      className,
+    ];
+
+    return dropdownClasses;
+  };
+
+  handleItemClick = (event) => {
+    /* TODO NO SURE IF GOOD WAY TO DO IT */
+    const { children } = this.props;
+    const index = event.target.id.split('_')[2];
+    if (index && (index < React.Children.count(children))) {
+      React.Children.toArray(children)[index].props.onClick();
+    }
+    event.stopPropagation(); // stopPropagation so we can have nested menus
+  };
+
+  renderChildren = () => {
+    const { children, checkbox } = this.props;
+    if (!children) return null;
+    return React.Children.map(children, (child, index) => {
+      const id = getUniqueHash('item', index);
+
+      if (child.type !== MenuItem) {
+        return React.cloneElement(
+          child, { key: id }
+        );
+      }
+      return React.cloneElement(
+        child, {
+          selected: checkbox ? child.props.selected === true : null,
+          // with this we set the selected prop to true or false (no undef or null)
+          // so that the child becomes a menuitemcheckbox (not menuitem)
+          key: id,
+          id: `${id}_${index}`,
+        }
+      );
+    });
+  };
+
+  render() {
+    const {
+      button,
+      height,
+      heightIcon,
+      isOpen,
+      renderClosedDropdown
+    } = this.props;
+
+    const listClasses = [
+      { [`slds-dropdown_length-${height}`]: height },
+      { [`slds-dropdown_length-with-icon-${heightIcon}`]: heightIcon },
+      'slds-dropdown__list',
+    ];
+
+    const rest = omit(this.props, Object.keys(propTypes));
+
+    return (
+      <div className={cx(this.getClasses())} {...rest}>
+        {button}
+        {(isOpen || renderClosedDropdown) && <div className={cx(this.getDropdownClasses())}>
+          <ul className={cx(listClasses)} role="menu" onClick={this.handleItemClick}>
```

Comment:
Why do we need the `onClick` at this point at all? The children `MenuItem` elements each have an `onClick`. Do I miss something?

Reply:
Reduce the number of event listeners. Apparently too many of those can slow down an app.

Reply:
@philippfrank The Menu re-work has two goals:

1. Improve the API, making it easier to use and extend in the future (SF has prototypes for sub menus now)
2. Speed up current features where a lot of Menus are used (MatchingConfiguration mounts ~25 menus with huge sub trees which slows rendering down significantly (seconds)

Binding one event listener here has the benefit that we do not have to loop over the child items and attach/detach event listeners when re-rendering. This can potentially be expensive when taking the new option `renderClosedDropdown` into account since we would have to rebind all event listeners every time the dropdown opens again. After implementing this, we can opt into performance optimizations in features such as MatchingConfig without negating the speed gains by binding/unbinding hundreds of EventListeners at once

Reply:
@fspoettel Makes sense. Nevertheless -- unless I'm missing something -- the MenuItems have their `onClick`-handler. The items are just cloned. Did you intend to remove them at that point? https://github.com/propertybase/react-lds/pull/293/files/00d5f8748a161eec579e96d7e4481dde6d0d3317#diff-e75026ec5544dc1bd4f13c5484b4ee42R132

Reply:
Yes. I now set ´onClick` to `null` when cloning.

Reply:
imo the best solution would be to attach an `onSelect` in `Menu` and call it with a hash`{ id, key }` of the MenuItem. This would free up `MenuItem` event listeners for custom logic. @brunottonurb unless I'm missing something after glancing over: How does nulling the `onClick` solve the problem when you still call it via `React.Children.toArray(children)[index].props.onClick();` in https://github.com/propertybase/react-lds/pull/293/files#diff-e75026ec5544dc1bd4f13c5484b4ee42R116?

Reply:
As far as I understand, the MenuItem `onClick` stays a prop, it's never rendered. So when rendering the MenuItems in ControlledMenu, I null the `onClick` prop, so it won't get rendered via `{...rest}` in MenuItem. ControlledMenu can still access the MenuItems `onClick`, because its props still contain the unchanged MenuItems.
I think this is more intuitive to use than an `onSelect`. When using the Menu, you don't even need to know we reduce the number of event listeners and pass your `onClick` to MenuItem like you'd always do.

Reply:
Well, what @fspoettel was saying is that the *attachment/detachment* of a huge number of listeners on the menu item level is the problem. Since you clone the original children, you also clone the `onClick` prop which results in an event listener to be attached for each menu item. 

When I get the plan correctly here, we want only a handler on the entire menu itself (whether we call that `onSelect` or something else; maybe `onItemSelect`?) instead of one on each item.

This results in a click on an item will be caught by the event listener which was attached to the menu which then reads the items ID and passes that as an argument to the `onSelect` handler. 

5 menus on a page with each 10 menu items will then result in 5 event listeners instead of 50.

Reply:
@brunottonurb 

> pass your onClick to MenuItem like you'd always do

There is nothing stopping you from doing that if we provide an `onSelect` callback. Attaching events to parent components is something you see a lot in bigger component libraries and it definitely makes sense when your component can contain hundreds of items and render frequently, making binding/unbinding events an expensive task. If we free up the event listener on `MenuItem` and let users set it on their own they can choose to bind `onClick` handlers to `MenuItem` and ignore that `onSelect` even exists. But having an `onSelect` and one instead of a couple hundred event listeners is something that should be possible.

Reply:
``` jsx
// Native Select in React
const select = (
  <select onChange={this.handle}>
    <option value="a">A</option>
    <option>B</option>
  </select>
);

// Suggested Menu API (Reference key is up for discussion, id would work, a data-attribute could be preferential. We could set that automatically when cloning the child)
const menu = (
  <Menu onSelect={this.handle}>
    <MenuItem id="a" data-value="a">A</MenuItem>
    <MenuItem id="b">b</MenuItem>
  </Menu>
);

// Still possible
const menuChildClick = (
  <Menu>
    <MenuItem onClick={this.handle}>A</MenuItem>
    <MenuItem onClick={this.handle}>B</MenuItem>
  </Menu>
```